kobotoolbox/kpi
# Refactor deployment project history logs

## Checklist

1. [x] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Run `./python-format.sh` to make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/main/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes
8. [x] Create a testing plan for the reviewer and add it to the Testing section

## Description

Changes `version_uid` in the project-history logs for re/deploy and un/archive to `latest_version_uid` and `latest_deployed_version_uid`. Otherwise developer-only changes.

## Notes

1. Moves the AuditAction choices to a new file since by the end of this project the list will be very long
2. Update AuditLoggedModelViewSet to add the log type to the request before doing anything else
3 . Adds middleware to look for a `log_type` attribute on a request and if `request.log_type == "project-history"`, call the `create_from_request` method in ProjectHistoryLog
4. Updates the AssetViewSet to inherit from AuditLoggedModelViewSet, tracking the `id` and `has_deployment` fields
5. Updates the `deployment` endpoint in the AssetViewSet to add a new `additional_audit_log_data` dictionary to the request with an `only_active_changed` field set to True if the request data was only {"active": <True/False>} (meaning the request was only an un/archive request). Also stores the value of the `active` param in the request and latest version and latest deployed version uids to put on the log.
  a. It only does this after the rest of the request succeeds, so we won't log failed attempts
7. Updates the code in `ProjectHistoryLog .create_from_deployment_request` to use the attributes put on the request instead of passed parameters. This is more in line with how the other logs will be created.
8.  If `additional_audit_log_data.only_active_changed` is true, sets the action to `archived` if `additional_audit_log_data.active` is False, `unarchived` otherwise.
9. If `additional_audit_log_data.only_active_changed` is missing or false, sets the action `deployed`  if `initial_data.has_deployment` is False, `redeployed` otherwise
10. Unit tests - Moves the project history log tests to their own file instead of sticking lines in test_api_asset since the code has been slightly more decoupled

## Testing

1. Create a new project 
2. Deploy the project
10. Go to `/api/v2/audit-logs/?q=log_type:project-history AND metadata__asset_uid:{project_uid}`
11. You should see the following project history log:
```
        {
            "app_label": "kpi",
            "model_name": "asset",
            "user": "http://kf.kobo.local:8080/api/v2/users/{you}/",
            "user_uid": "{your uid}",
            "username": "{your username}",
            "action": "deploy",
            "metadata": {
                "source": "{browser} ({OS})",
                "asset_uid": "{new project uid}",
                "ip_address": "172.18.0.1",
                "log_subtype": "project",
                "latest_version_uid": "{version uid}",
                "latest_deployed_version_uid": "{version uid}"
            },
            "date_created": "{whenever}",
            "log_type": "project-history"
        }
```
12. You can verify that the version id is correct by looking in the django shell for `Asset.objects.filter(uid=<uid>).latest_version.uid`
13. Redeploy the asset
14. Reload the endpoint and you should see a new log (the latest version and latest deployed version uid should be the same)
```
        {
            "app_label": "kpi",
            "model_name": "asset",
            "user": "http://kf.kobo.local:8080/api/v2/users/{you}/",
            "user_uid": "{your uid}",
            "username": "{your username}",
            "action": "redeploy",
            "metadata": {
                "source": "{browser} ({OS})",
                "asset_uid": "{new project uid}",
                "ip_address": "172.18.0.1",
                "log_subtype": "project",
                "latest_version_uid": "{version uid}",
                "latest_deployed_version_uid": "{version uid}"
            },
            "date_created": "{whenever}",
            "log_type": "project-history"
        }
```
15. Make a change to the project form and redeploy
16. You should see a new log with new version uids in the metadata
17. Go to Settings > General and archive the project
18. Reload the endpoint and you should see a new log with action: "archive" and the latest_deployed_version_uid in the metadata (as well as the rest of the usual data and metadata)
19. Unarchive the the project
20. Reload the endpoint and you should see a new log with action:"unarchive" and the latest_deployed_version_uid
21. Go to the projects table and select the new project
22. Hit the archive icon in the corner
23. Verify a new archive log shows up
24. Hit the unarchive icon
25. Verify a new unarchive log shows up
26. In your terminal, run `curl -X POST -H "Authorization: Token <your token>" -H 'Content-Type: application/json' -H 'Accept: application/json' http://kf.kobo.local:8080/api/v2/assets/<asset_uid>/deployment/ --data '{"active":"false"}'`
27. The request should fail and no log should be generated
28. Run the same request with PATCH instead of POST
29. The request should succeed and a new audit log with action 'archive' should be generated
30. 19. In your terminal, run `curl -X POST -H "Authorization: Token <your token>" -H 'Content-Type: application/json' -H 'Accept: application/json' http://kf.kobo.local:8080/api/v2/assets/<asset_uid>/deployment/ --data '{"active":"false", "backend":"openrosa"}'`
20. The request should fail and no log should be generated
21. Run the same request with PATCH instead of POST
22. The request should succeed and a new audit log with action 'redeploy' should be generated
